File: src/main/java/cc/sukazyo/cono/morny/daemon/MedicationTimer.java
package cc.sukazyo.cono.morny.daemon;

import cc.sukazyo.cono.morny.MornyCoeur;
import cc.sukazyo.cono.morny.util.CommonFormat;
import com.pengrad.telegrambot.model.Message;
import com.pengrad.telegrambot.model.MessageEntity;
import com.pengrad.telegrambot.model.request.ParseMode;
import com.pengrad.telegrambot.request.EditMessageText;
import com.pengrad.telegrambot.request.SendMessage;
import com.pengrad.telegrambot.response.SendResponse;

import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import static cc.sukazyo.cono.morny.Log.exceptionLog;
import static cc.sukazyo.cono.morny.Log.logger;

public class MedicationTimer extends Thread {
	
	public static final ZoneOffset USE_TIME_ZONE = ZoneOffset.ofHours(8);
	public static final Set<Integer> NOTIFY_AT_HOUR = Set.of(12);
	public static final long NOTIFY_CHAT = -1001729016815L;
	public static final String NOTIFY_MESSAGE = "\uD83C\uDF65⏲";
	private static final String DAEMON_THREAD_NAME = "TIMER_Medication";
	
	private static final long LAST_NOTIFY_ID_NULL = -1L;
	private long lastNotify = LAST_NOTIFY_ID_NULL;
	
	
	MedicationTimer () {
		super(DAEMON_THREAD_NAME);
	}
	
	@Override
	public void run () {
		logger.info("MedicationTimer started");
		while (!interrupted()) {
			try {
				waitToNextRoutine();
				sendNotification();
			} catch (InterruptedException e) {
				interrupt();
				logger.info("MedicationTimer was interrupted, will be exit now");
			} catch (Exception e) {
				logger.error("Unexpected error occurred");
				logger.error(exceptionLog(e));
				MornyReport.exception(e);
			}
		}
		logger.info("MedicationTimer stopped");
	}
	
	private void sendNotification () {
		final SendResponse resp = MornyCoeur.extra().exec(new SendMessage(NOTIFY_CHAT, NOTIFY_MESSAGE));
		if (resp.isOk()) lastNotify = resp.message().messageId();
		else lastNotify = LAST_NOTIFY_ID_NULL;
	}
	
	public void refreshNotificationWrite (Message edited) {
		if (edited.messageId() != lastNotify) return;
		final String editTime = CommonFormat.formatDate(edited.editDate()*1000, 8);
		ArrayList<MessageEntity> entities = new ArrayList<>();
		if (edited.entities() != null) entities.addAll(List.of(edited.entities()));
		entities.add(new MessageEntity(MessageEntity.Type.italic, edited.text().length() + "\n-- ".length(), editTime.length()));
		EditMessageText sending = new EditMessageText(
				NOTIFY_CHAT,
				edited.messageId(),
				edited.text() + "\n-- " + editTime + " --"
		).parseMode(ParseMode.HTML).entities(entities.toArray(MessageEntity[]::new));
		MornyCoeur.extra().exec(sending);
		lastNotify = LAST_NOTIFY_ID_NULL;
	}
	
	public static long calcNextRoutineTimestamp (long baseTimeMillis, ZoneOffset useTimeZone, Set<Integer> atHours) {
		LocalDateTime time = LocalDateTime.ofEpochSecond(
				baseTimeMillis/1000, (int)baseTimeMillis%1000*1000,
				useTimeZone
		).withMinute(0).withSecond(0).withNano(0);
		do {
			time = time.plusHours(1);
		} while (!atHours.contains(time.getHour()));
		return time.withMinute(0).withSecond(0).withNano(0).toInstant(useTimeZone).toEpochMilli();
	}
	
	private void waitToNextRoutine () throws InterruptedException {
		sleep(calcNextRoutineTimestamp(System.currentTimeMillis(), USE_TIME_ZONE, NOTIFY_AT_HOUR) - System.currentTimeMillis());
	}
	
}

File: src/main/scala/cc/sukazyo/cono/morny/daemon/MedicationTimer.scala
package cc.sukazyo.cono.morny.daemon

import cc.sukazyo.cono.morny.Log.logger
import cc.sukazyo.cono.morny.MornyCoeur
import cc.sukazyo.cono.morny.daemon.MedicationTimer.calcNextRoutineTimestamp
import cc.sukazyo.cono.morny.util.schedule.RoutineTask
import cc.sukazyo.cono.morny.util.tgapi.TelegramExtensions.Bot.exec
import cc.sukazyo.cono.morny.util.CommonFormat
import com.pengrad.telegrambot.model.{Message, MessageEntity}
import com.pengrad.telegrambot.request.{EditMessageText, SendMessage}
import com.pengrad.telegrambot.response.SendResponse

import java.time.{LocalDateTime, ZoneOffset}
import scala.collection.mutable.ArrayBuffer
import scala.language.implicitConversions

class MedicationTimer (using coeur: MornyCoeur) {
	
	private val NOTIFY_MESSAGE = "🍥⏲"
	private val DAEMON_THREAD_NAME_DEF = "MedicationTimer"
	
	private val use_timeZone = coeur.config.medicationTimerUseTimezone
	import scala.jdk.CollectionConverters.SetHasAsScala
	private val notify_atHour: Set[Int] = coeur.config.medicationNotifyAt.asScala.toSet.map(_.intValue)
	private val notify_toChat = coeur.config.medicationNotifyToChat
	
	private var lastNotify_messageId: Option[Int] = None
	
	private val scheduleTask: RoutineTask = new RoutineTask {
		
		override def name: String = DAEMON_THREAD_NAME_DEF
		
		def calcNextSendTime: Long =
			val next_time = calcNextRoutineTimestamp(System.currentTimeMillis, use_timeZone, notify_atHour)
			logger info s"medication timer will send next notify at ${CommonFormat.formatDate(next_time, use_timeZone.getTotalSeconds / 60 / 60)} with $use_timeZone [$next_time]"
			next_time
		
		override def firstRoutineTimeMillis: Long =
			calcNextSendTime
		
		override def nextRoutineTimeMillis (previousRoutineScheduledTimeMillis: Long): Long | Null =
			calcNextSendTime
		
		override def main: Unit = {
			sendNotification()
			logger info "medication notify sent."
		}
		
	}
	
	def start(): Unit =
		if ((notify_toChat == -1) || (notify_atHour isEmpty))
			logger notice "Medication Timer disabled : related param is not complete set"
			return;
		coeur.tasks ++ scheduleTask
		logger notice "Medication Timer started."
	
	def stop(): Unit =
		coeur.tasks % scheduleTask
		logger notice "Medication Timer stopped."
	
	private def sendNotification(): Unit = {
		val sendResponse: SendResponse = coeur.account exec SendMessage(notify_toChat, NOTIFY_MESSAGE)
		if sendResponse isOk then lastNotify_messageId = Some(sendResponse.message.messageId)
		else lastNotify_messageId = None
	}
	
	def refreshNotificationWrite (edited: Message): Unit = {
		if (lastNotify_messageId isEmpty) || (lastNotify_messageId.get != (edited.messageId toInt)) then return
		import cc.sukazyo.cono.morny.util.CommonFormat.formatDate
		val editTime = formatDate(edited.editDate*1000, use_timeZone.getTotalSeconds/60/60)
		val entities = ArrayBuffer.empty[MessageEntity]
		if edited.entities ne null then entities ++= edited.entities
		entities += MessageEntity(MessageEntity.Type.italic, edited.text.length + "\n-- ".length, editTime.length)
		coeur.account exec EditMessageText(
			notify_toChat,
			edited.messageId,
			edited.text + s"\n-- $editTime --"
		).entities(entities toArray:_*)
		lastNotify_messageId = None
	}
	
}

object MedicationTimer {
	
	@throws[IllegalArgumentException]
	def calcNextRoutineTimestamp (baseTimeMillis: Long, zone: ZoneOffset, notifyAt: Set[Int]): Long = {
		if (notifyAt isEmpty) throw new IllegalArgumentException("notify time is not set")
		var time = LocalDateTime.ofEpochSecond(
			baseTimeMillis / 1000, ((baseTimeMillis % 1000) * 1000 * 1000) toInt,
			zone
		).withMinute(0).withSecond(0).withNano(0)
		time = time plusHours 1
		while (!(notifyAt contains(time getHour))) {
			time = time plusHours 1
		}
		(time toInstant zone) toEpochMilli
	}
	
}
